### ISSUE TITLE
Relative paths in `CMAKE_INSTALL_PREFIX`

### ISSUE THREAD

**Issue by Meakk** [google/draco#955]
In `draco_install.cmake`, `CMAKE_INSTALL_FULL_*` are used to define the destination for the different targets.  

https://github.com/google/draco/blob/7ec8a2783f4c86d5f3c9120d1d58c7ce1b3094ef/cmake/draco_install.cmake#L26

However, when using these variables, `CMAKE_INSTALL_PREFIX` is effectively prepended twice when using a relative path.  
Here is the output of my installation step when using `CMAKE_INSTALL_PREFIX` set to `./install`:

```
[...]
-- Installing: /home/michael/dev/draco/build/./install/./install/include/draco/draco_features.h
-- Installing: /home/michael/dev/draco/build/./install/./install/bin/draco_decoder-1.5.5
-- Installing: /home/michael/dev/draco/build/./install/./install/bin/draco_decoder
-- Installing: /home/michael/dev/draco/build/./install/./install/bin/draco_encoder-1.5.5
-- Installing: /home/michael/dev/draco/build/./install/./install/bin/draco_encoder
-- Installing: /home/michael/dev/draco/build/./install/./install/lib/libdraco.a
-- Installing: /home/michael/dev/draco/build/./install/./install/lib/pkgconfig/draco.pc
-- Installing: /home/michael/dev/draco/build/./install/./install/share/cmake/draco/draco-targets.cmake
-- Installing: /home/michael/dev/draco/build/./install/./install/share/cmake/draco/draco-targets-release.cmake
-- Installing: /home/michael/dev/draco/build/./install/./install/share/cmake/draco/draco-config.cmake
-- Installing: /home/michael/dev/draco/build/./install/./install/share/cmake/draco/draco-config-version.cmake
```

Moreover, it's causing a `find_package` failure.  
In my case I'm calling `-Ddraco_DIR=/home/michael/dev/draco/build/install/install/share/cmake/draco` (note the `install` folder specified twice) and I got the following error:
```
CMake Error at
 /home/michael/dev/draco/build/install/install/share/cmake/draco/drac
 o-targets.cmake:83 (message):
   The imported target "draco::draco" references the file

      "/home/michael/dev/draco/build/./install/lib/libdraco.a"

   but this file does not exist.  Possible reasons include:

   * The file was deleted, renamed, or moved to another location.

   * An install or uninstall procedure did not complete
 successfully.

   * The installation package was faulty and contained


 "/home/michael/dev/draco/build/install/install/share/cmake/draco/dra
 co-targets.cmake"

   but not all the files it references.

```

As you can see, the `install` folder for `libdraco.a` is specified only once resulting in this failure.  
Changing all the `CMAKE_INSTALL_FULL_*` to `CMAKE_INSTALL_*` fixes the problem.

My CMake version is 3.25.0 by the way but it's probably not relevant.

**Comment by Meakk** [comment]
Indeed, if you pass the option directly to the CMake command, it works because CMake is expanding the relative path to an absolute path before the configuration.  

If you just run `cmake ..`, then edit `CMAKE_INSTALL_PREFIX` with `ccmake` or `cmake-gui`, the relative path stays relative and is duplicated during the installation step.

**Comment by Meakk** [comment]
I also notice that if you add `:PATH` to the variable, it is also duplicating the prefix:

```sh
cmake .. -DCMAKE_INSTALL_PREFIX:PATH=./install_root && make -j install
```

**Comment by tomfinegan** [comment]
> If you just run cmake .., then edit CMAKE_INSTALL_PREFIX with ccmake or cmake-gui, the relative path stays relative and is duplicated during the installation step.

I don't understand why you expect adding a prefix a second time is going to result in CMake understanding that you want the prefix to appear only once in the install path. 

1) CMake generates the build with the original prefix.
2) You specify an additional prefix.
3) CMake prefixes all the install paths generated in step 1.

The above sounds like CMake is working as intended, and I don't see how using the relative forms of the install target vars will help here: you're still going to get the additional relative prefix when you rerun the tool-- the vars have already been expanded by the second run.

If you want to change the install prefix I suggest that you regenerate the build after wiping out the configuration directory.

**Comment by Meakk** [comment]
Modifying the value in ccmake or cmake-gui does not append the prefix, it is supposed to replace the value.
Look at my first log in the initial post. I have set `./install` in ccmake only once, but it appears twice.  

Also, setting it using by specifying the type in the command line does not work either:

```
cmake .. -DCMAKE_INSTALL_PREFIX:PATH=./install_root && make -j install
```

Setting `CMAKE_INSTALL_PREFIX` (and all other options) after the initial configuration is a quite standard CMake usage.

**Comment by tomfinegan** [comment]
As I said before I don't understand why you expect this to work. 

Draco's CMake install target uses the CMake provided full path variants of the GNUInstallDirs variables, and attempting to change the prefix in a build that's already had its variables expanded to full paths is obviously going to break by prefixing the existing prefix. 

In other words, it's clearly adding a prefix to a prefix, and prefixing aka _prepending_ them to the already expanded path.

If you have a patch that passes your test case and does not break the existing Draco installation tests I'll be happy to review a PR.

**Comment by tomfinegan** [comment]
Can you explain what you're actually trying to do? 

If you're simply trying to install an existing build to a different location why not use DESTDIR?

`make DESTDIR=<install path> install`

**Comment by Meakk** [comment]
I'm integrating draco in [F3D](https://f3d.app) to support drc files, and later gltf draco extension. 

I noticed the install issue locally by using ccmake. I ignored it and tried integrating draco in my CI using `-DCMAKE_INSTALL_PREFIX:PATH`, and I observed the same problem, so I provided a patch.  
You can take a look at the initial commit here: https://github.com/f3d-app/f3d/commit/f4ef7a1c278f08cbd6901a8523798f1d390eefea

Note that I'm using the typed option in all other dependencies we have without any problem.
The commit also contains a patch to fix the wrong install path.

Yesterday, I added a new commit to remove the typed option since it works fine as you pointed out (I also forced PIC but it's unrelated to this issue)
The commit is here: https://github.com/f3d-app/f3d/commit/c1fc72573bdff4b3d595c7fcca802f65556cd81a

Anyway, I managed to have a successful installation, but I still think there is an issue on your side.

**Comment by tomfinegan** [comment]
I was looking at the last patch, and this line stands out: https://github.com/f3d-app/f3d/commit/c1fc72573bdff4b3d595c7fcca802f65556cd81a#diff-273d72774610d16361b37ea58e95cb9e192f2991e443a84576253b24031c317aL53

The removal of `:PATH` from the step setting CMAKE_INSTALL_PREFIX in your CMake config command is what addresses the issue here. I am able to reproduce your initial issue by adding that to my command line, so I agree that something is wrong here. The difference in `CMakeCache.txt` is curious here.

```
# with :PATH the install path is wrong.
CMAKE_INSTALL_PREFIX:PATH=./install

-- Installing: /Users/tomfinegan/_werk/draco/.build_install_test/./install/./install/lib/libdraco.a
```

```
# without :PATH it's fine.
CMAKE_INSTALL_PREFIX:PATH=/Users/tomfinegan/_werk/draco/.build_install_test/install

-- Installing: /Users/tomfinegan/_werk/draco/.build_install_test/install/lib/libdraco.a
```

This is with a single generation and build, and not multiple CMake runs in a single generation/build directory. I'll look more deeply into this, but this is very strange behavior.

**Comment by Meakk** [comment]
> but that the CMake Unix Makefiles generator is doing something very strange

I forgot to mention, but I reproduce with Ninja

Thanks for the detailed analysis
